text/0000-build-std.md

```diff
@@ -0,0 +1,3601 @@
+components, such as `rust-mingw`, rustc's `-Clink-self-contained` will be able
+to link against the object files and build-std should never fail on account of
+missing special object files.
+
+### `compiler-builtins-mem`
+[compiler-builtins-mem]: #compiler-builtins-mem
+
+The `mem` feature of `compiler_builtins` (and the subsequent
+`compiler-builtins-mem` feature of `core`, `alloc`, `std` which forward to
+`compiler_builtins/mem`) is required by `no_std` crates because `libc` is not
+providing these symbols without `std`.
+
+It is necessary that the `compiler-builtins-mem` feature of `alloc` and/or
+`core` be enabled when `std` is not in the crate graph.
+
+*See the following sections for rationale/alternatives:*
+- [*Why not use weak linkage for `compiler-builtins/mem` symbols?*][why-not-use-weak-linkage-for-compiler-builtins-mem-symbols]
+
+### Potential migration breakage
+[potential-migration-breakage]: #potential-migration-breakage
+
+When building an existing `no_std` project for a tier three target with
+build-std, there could be an implicit dependency on `std` from a dependency
+`no_std` crate that has not yet made its dependency on only the `core` crate
+explicit, for example. In this circumstance, this would fail to build as the
+target will have had `standard_library_support.std = false` in its target
+specification and Cargo will refuse to build `std` (see
+[*Support of the standard library for a target*][support-of-the-standard-library-for-a-target]).
+
+*See the following sections for rationale/alternatives:*
+- [*Why permit breakage of nightly build-std users using tier three targets?*][why-permit-breakage-of-nightly-build-std-users-using-tier-three-targets]
+
+### Caching
+[caching]: #caching
+
+Standard library artifacts build by built-std will not be shared between crates
+or workspaces, as they only exist in the `target` directory of a specific crate
+or workspace.
+
+*See the following sections for rationale/alternatives:*
+- [*Why not globally cache builds of the standard library?*][why-not-globally-cache-builds-of-the-standard-library]
+
+### Sanitizers
+[sanitizers]: #sanitizers
+
+rustc's sanitizer support is currently unstable as it is not possible for users
+to re-build the standard library with sanitizer support.
+
+It is out-of-scope for this RFC to propose stabilising sanitizers (see
+[rust#123617]) or to expose sanitizer configuration in Cargo, but it is
+instructive to examine how build-std would enable sanitizer support to ensure
+that the proposed design is compatible.
+
+rustc's flag to enable sanitizers will be a target modifier, as the
+instrumentation must be present for all of the crates to avoid false positives.
+
+rustc's sanitizer support attempts to locate sanitizer runtimes in the sysroot
+(`$sysroot/lib/rustlib/$target/lib/`) to link against. Rust already ships
+sanitizer runtimes for targets that support sanitizers, so all that is necessary
```

Comment:
How do these runtimes interact with other target modifiers like CFI?

Reply:
Discussed in person - sanitizers are generally development tools while CFI is mainly useful in shipped binaries, so it seems reasonable to disallow their use together. Other target modifiers exist, and strictly speaking this isn't a build-std issue, but it might be worth making a note that this isn't everything necessary to allow mixing all target modifiers because build-std cannot rebuild all builtin dependencies.